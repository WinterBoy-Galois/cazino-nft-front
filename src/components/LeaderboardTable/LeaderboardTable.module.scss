@import '../../styles/shared';

$radius: 10px;

.leaderboard-table {
  min-width: 100%;
  color: $white;
  border-spacing: 0;
  border-collapse: collapse;

  td {
    padding: 0;
  }

  &__header {
    th {
      position: sticky;
      top: 0;
      font-weight: 300;
      padding: 0.5rem;
      background-color: $table-light-blue;
      text-transform: uppercase;
      color: $table-ultra-light-blue;
      font-size: 15px;
      text-align: left;
      z-index: 1;
    }

    th:first-child {
      // Fixes Safari iOS border radius
      @supports (-webkit-touch-callout: none) {
        border-radius: $radius 0 0 0;
      }

      padding-left: 1rem;
    }

    th:last-child {
      // Fixes Safari iOS border radius
      @supports (-webkit-touch-callout: none) {
        border-radius: 0 $radius 0 0;
      }

      padding-right: 1rem;
    }
  }

  &__body {
    background-color: $table-dark-blue;
    font-size: 13px;

    @include media-breakpoint-up(xs) {
      font-size: 13px;
    }

    td div {
      padding: 0.5rem 0.5rem;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      max-width: 130px;

      @include media-breakpoint-up(lg) {
        max-width: 129px;
        padding: 0.6rem 0.5rem;
      }
    }

    tr td:first-child div {
      padding-left: 1rem;
    }

    tr td:last-child div {
      padding-right: 1rem;
    }
  }

  &__wrapper {
    max-height: calc(100% - 0.5rem);
    min-height: 200px;
    height: auto;
    width: 100%;
    overflow-x: hidden;
    overflow-y: auto;
    white-space: nowrap;
    background-color: $table-dark-blue;
    border-radius: $radius;
    display: flex;
    flex-direction: column;

    @include media-breakpoint-up(md) {
      overflow-y: hidden;
    }

    @include media-breakpoint-up(lg) {
      max-height: calc(100% - 1.5rem);
    }
  }
}

.fade-enter {
  opacity: 0;
  transform: translateX(-50%);
}

.fade-enter-active {
  opacity: 1;
  transform: translateX(0%);
}

.fade-exit {
  opacity: 1;
  transform: translateX(0%);
}

.fade-exit-active {
  opacity: 0;
  transform: translateX(50%);
}

.fade-enter-active,
.fade-exit-active {
  transition: opacity 500ms, transform 500ms;
}
