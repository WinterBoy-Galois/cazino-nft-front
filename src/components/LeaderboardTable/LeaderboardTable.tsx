import React from 'react';
import styles from './LeaderboardTable.module.scss';
import { Leader } from '../../models/leader.model';
import LeaderboardRow from './components/LeaderboardRow';
import { useBreakpoint } from '../../hooks/useBreakpoint.hook';
import { SwitchTransition, CSSTransition } from 'react-transition-group';
import SpacerRow from './components/SpacerRow';
import Loading from '../Loading';
import Error from '../Error';
import { useTranslation } from 'react-i18next';

interface IProps {
  leaderboard: Leader[];
  isLoading: boolean;
  error: boolean;
  signInUserId: string;
  onRowClicked?: () => void;
  onUsernameClicked?: (userId: string) => void;
}

const LeaderboardTable: React.FC<IProps> = ({
  leaderboard,
  isLoading,
  error,
  signInUserId,
  onRowClicked,
  onUsernameClicked,
}) => {
  const breakpoint = useBreakpoint();
  const { t } = useTranslation(['sidebar']);

  const renderBonusColumn = () => {
    switch (breakpoint) {
      case 'xs':
      case 'sm':
      case 'md':
        return false;
      default:
        return true;
    }
  };

  return (
    <div className={styles['leaderboard-table__wrapper']}>
      <table className={styles['leaderboard-table']}>
        <thead className={styles['leaderboard-table__header']}>
          <tr>
            <th>#</th>
            <th>{t('leaderboards.table.user')}</th>
            <th>{t('leaderboards.table.wagered')}</th>
            {renderBonusColumn() && <th>{t('leaderboards.table.bonus')}</th>}
          </tr>
        </thead>
        <tbody className={styles['leaderboard-table__body']}>
          {leaderboard && leaderboard.length > 0 ? (
            <>
              <SpacerRow />
              {leaderboard.map((l: Leader, i) => (
                <SwitchTransition key={i} mode="out-in">
                  <CSSTransition
                    key={`${l.userid}`}
                    classNames={{
                      enter: styles['fade-enter'],
                      enterActive: styles['fade-enter-active'],
                      exit: styles['fade-exit'],
                      exitActive: styles['fade-exit-active'],
                    }}
                    timeout={500}
                  >
                    <LeaderboardRow
                      leader={l}
                      place={i + 1}
                      highlight={l.userid === signInUserId}
                      onRowClicked={onRowClicked}
                      onUsernameClicked={onUsernameClicked}
                    />
                  </CSSTransition>
                </SwitchTransition>
              ))}
              <SpacerRow />
            </>
          ) : null}
        </tbody>
      </table>
      {!error && isLoading && (leaderboard.length <= 0 || !leaderboard) && <Loading />}
      {error && !isLoading && (leaderboard.length <= 0 || !leaderboard) && (
        <Error>Sorry, there was an unexpected error.</Error>
      )}
      {!error && !isLoading && (leaderboard.length <= 0 || !leaderboard) && (
        <Error>Sorry, received no data.</Error>
      )}
    </div>
  );
};

export default LeaderboardTable;
