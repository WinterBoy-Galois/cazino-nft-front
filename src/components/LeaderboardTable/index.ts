export { default } from './LeaderboardTable';
export { default as Error } from './components/Error';
export { default as Empty } from './components/Empty';
export { default as Loading } from './components/Loading';
