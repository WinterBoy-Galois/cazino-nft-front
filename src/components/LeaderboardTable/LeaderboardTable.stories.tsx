import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import LeaderboardTable from '.';
import { shuffle } from '../../common/util';
import { action } from '@storybook/addon-actions';

const data = [
  {
    username: 'dailychampion1',
    userid: '1',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion2',
    userid: '2',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion3',
    userid: '3',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion4',
    userid: '4',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion5',
    userid: '5',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion6',
    userid: '6',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion7',
    userid: '7',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion8',
    userid: '8',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion7',
    userid: '9',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
  {
    username: 'dailychampion8',
    userid: '10',
    wager: 0.12345678,
    bonus: 0.321321401,
  },
];

const LeaderboardTableWrapper = () => {
  const [leaderboard, setLeaderboard] = useState(data);

  return (
    <>
      <LeaderboardTable
        error={false}
        isLoading={false}
        leaderboard={leaderboard}
        signInUserId="5"
      />
      <button onClick={() => setLeaderboard(shuffle(data))}>shuffle</button>
    </>
  );
};

storiesOf('Components/LeaderboardTable', module)
  .add('default', () => (
    <LeaderboardTable
      error={false}
      isLoading={false}
      leaderboard={data}
      signInUserId="5"
      onRowClicked={action('Row clicked')}
      onUsernameClicked={action('Username clicked')}
    />
  ))
  .add('shuffle data', () => <LeaderboardTableWrapper />)
  .add('empty', () => (
    <LeaderboardTable
      error={false}
      isLoading={false}
      leaderboard={[]}
      signInUserId="5"
      onRowClicked={action('Row clicked')}
      onUsernameClicked={action('Username clicked')}
    />
  ))
  .add('loading', () => (
    <LeaderboardTable error={false} isLoading={true} leaderboard={[]} signInUserId="5" />
  ))
  .add('error', () => (
    <LeaderboardTable error={true} isLoading={false} leaderboard={[]} signInUserId="5" />
  ))
  .add('signed in user not in leaderboard', () => (
    <LeaderboardTable
      error={false}
      isLoading={false}
      leaderboard={data}
      signInUserId="-1"
      onRowClicked={action('Row clicked')}
      onUsernameClicked={action('Username clicked')}
    />
  ))
  .add('scrolling', () => (
    <div style={{ height: '250px' }}>
      <LeaderboardTable
        error={false}
        isLoading={false}
        leaderboard={data}
        signInUserId="5"
        onRowClicked={action('Row clicked')}
        onUsernameClicked={action('Username clicked')}
      />
    </div>
  ))
  .addDecorator(storyFn => (
    <div className="container" style={{ height: '500px' }}>
      {storyFn()}
    </div>
  ));
