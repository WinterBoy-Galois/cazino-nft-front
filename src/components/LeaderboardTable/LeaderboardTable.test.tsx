import React from 'react';
import { render } from '@testing-library/react';
import LeaderboardTable from '.';

test('renders', async () => {
  const { container } = render(
    <LeaderboardTable leaderboard={[]} isLoading={false} error={false} signInUserId={'1'} />
  );
  expect(container).toMatchSnapshot();
});
