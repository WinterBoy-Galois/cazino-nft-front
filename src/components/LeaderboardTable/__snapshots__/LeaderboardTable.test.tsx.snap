// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="leaderboard-table__wrapper"
  >
    <table
      class="leaderboard-table"
    >
      <thead
        class="leaderboard-table__header"
      >
        <tr>
          <th>
            #
          </th>
          <th>
            User
          </th>
          <th>
            Wagered
          </th>
          <th>
            Bonus
          </th>
        </tr>
      </thead>
      <tbody
        class="leaderboard-table__body"
      />
    </table>
    <div
      class="undefined container"
    >
      Sorry, received no data.
    </div>
  </div>
</div>
`;
